File: solutions/go/largest-series-product/1/largest_series_product.go
package lsproduct

import (
	"errors"
	"unicode"
)

func LargestSeriesProduct(digits string, size int) (int64, error) {
	if size < 0 {
		return 0, errors.New("span must be positive or zero")
	}
	if size > len(digits) {
		return 0,
			errors.New(
				"span must be less or equal to string length",
			)
	}
	if size == 0 {
		return 1, nil
	}

	maxProduct := int64(0)
	for i := 0; i <= len(digits)-size; i++ {
		product := int64(1)
		for j := range size {
			digit := digits[i+j]
			if !unicode.IsDigit(rune(digit)) {
				return 0,
					errors.New(
						"input string must only contain digits",
					)
			}
			product *= int64(digit - '0')
		}
		if product > maxProduct {
			maxProduct = product
		}
	}
	return maxProduct, nil
}

File: solutions/c/largest-series-product/1/largest_series_product.c
#include "largest_series_product.h"

#include <ctype.h>
#include <string.h>

int64_t largest_series_product(char* digits, size_t span) {
    size_t digits_len = strlen(digits);
    if (span > digits_len) return -1;
    int64_t max_product = 0;
    for (size_t i = 0; i <= digits_len - span; i++) {
        int64_t product = 1;
        for (size_t j = i; j < i + span; j++) {
            if (isdigit(digits[j]))
                product *= digits[j] - '0';
            else
                return -1;
        }
        if (product > max_product) max_product = product;
    }
    return max_product;
}
